### PR TITLE
DB-9108 allow subquery in the ON clause of flattened inner join node …

### PR DESCRIPTION
**Pull request by yxia92**
…to be flattened

Also updated two test cases in ExportFileTest.java and ExportPermissionCheckTest.java to make IT run on macOS Catalina.

### CODE REVIEW

**Comment by yxia92** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by yxia92** [comment]
jenkins please retest branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by yxia92** [comment]
jenkins please retest branch @cdh5.12.2

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportFileTest.java [hunk 1]
```diff
@@ -111,7 +111,8 @@ public void createDirectory_returnsFalseWhenCannotCreate() throws IOException, S
             assertFalse(exportFile.createDirectory());
         }
         catch (Exception e) {
-             Assert.assertTrue(e.getMessage(), e.getMessage().contains("IOException '/noPermissionToCreateFolderInRoot' when accessing directory"));
```
Comment (arnaud-splice) [inline splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportFileTest.java]: That's a good use case for hamcrest: more concise, more readable and out of the box comprehensive error message:
`assertThat(e.getMessage(), anyOf(containsString(strA), containsString(strB))`
or, if you're adventurous:
`assertThat(e.getMessage(), matchesPattern("IOException '/noPermissionToCreateFolderInRoot(: Read-only file system)+' when accessing directory"))`
The second one might need adjustments, I'm a bit rusty on patterns.

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportPermissionCheckTest.java [hunk 2]
```diff
@@ -47,12 +48,15 @@ public static void setupConfig() {
...
-        expectedException.expect(StandardException.class);
-        expectedException.expectMessage("IOException '/ExportPermissionCheckTest' when accessing directory");
-
         ExportParams exportParams = ExportParams.withDirectory("/ExportPermissionCheckTest");
         ExportPermissionCheck permissionCheck = new ExportPermissionCheck(exportParams,dfs);
-        permissionCheck.verify();
+        try {
+            permissionCheck.verify();
+        } catch (Exception e) {
+            Assert.assertTrue(e.getMessage(), e.getMessage().contains("IOException '/ExportPermissionCheckTest' when accessing directory")
```
Comment (arnaud-splice) [inline splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportPermissionCheckTest.java]: same

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HalfOuterJoinNode.java [hunk 3]
```diff
@@ -923,6 +923,11 @@ public FromList flatten(ResultColumnList rcl,
             havingClause.remapColumnReferencesToExpressions();
         }
 
+        if (delayPreprocessingJoinClause()) {
```
Comment (jyuanca) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HalfOuterJoinNode.java]: Does delayPreproessingJoinClause always return false for HalfOuterJoinNode?

Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HalfOuterJoinNode.java]: Yes, currently we don't have the logic to flatten ON clause subquery for outer joins (both Half and FullOuterJoinNode), so there is no need to delayPreprocessingJoinClause for Half and FullOuterJoinNode.

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportFileTest.java (hunk 1 above)
Comment (yxia92) [reply splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportFileTest.java]: done, thanks for the suggestion.

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportPermissionCheckTest.java (hunk 2 above)
Comment (yxia92) [reply splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/export/ExportPermissionCheckTest.java]: done.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HalfOuterJoinNode.java (hunk 3 above)
Comment (jyuanca) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HalfOuterJoinNode.java]: If it is always false, this block will never executed.

Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HalfOuterJoinNode.java]: Hi @jyuanca You are correct, this block is not exercised now. I put it here in anticipation that in the future we may support the flattening of ON clause subquery for outer joins too.

**Comment by spliceci** [comment]
TEST FAILED

**Comment by yxia92** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**